parallel/task_pool.go

```diff
@@ -0,0 +1,58 @@
+package parallel
+
+import (
+	"math"
+	"sync"
+)
+
+type job[T any] struct {
+	val T
+	i   int
+}
+
+func NewTaskPool[T any](source []T, poolSize int, fn func(T, int)) {
```

Comment:
Good implementation

I think we should add the following code on top of your function:

```go
if poolSize > len(source) {
   poolSize = len(source)
}
```

It will limit pool size in cases where source collection is smaller than the number of workers.

Reply:
Done